### ISSUE TITLE
What is causing AHC/2.1 to create mass amounts of HTTP connections?

### ISSUE THREAD

**Issue by fredsted** [AsyncHttpClient/async-http-client#1823]
Hello

I'm the founder of [Webhook.site](https://webhook.site), a tool for testing webhooks and HTTP requests. 

I've found that a lot of users connecting with the `AHC/2.1` user agent seem to create mass amounts of HTTP connections (seen with `netstat`), in the order of thousands, or tens of thousands, per IP address. These connections stay open for a while. It seems to be happening for different users, so it's not a single mis-configured user that this is happening with.

So far I've resorted to blocking the `AHC/2.1` user agent and manually blocking IP addresses and ranges of users who're using this software. But since this software seems to be a common way to send HTTP requests I'd like to know if there's anything I can do to avoid this on my side.

One thing of note is that we automatically block users sending large amounts of traffic with either a 410 or a 429 HTTP error. Could this be what's causing it? Does AHC/2.1 have some sort of auto-retry mechanism enabled per default?

**Comment by fredsted** [comment]
Thanks for the quick response @slandelle 

One thing I'm noting that this issue is specific to the AHC/2.1 software. I'm not seeing this issue on any other HTTP clients. Webhook.site is used by hundreds of thousands of users every month and our customers are using all sorts of HTTP clients.

I don't believe that this is caused by ill intentions, but rather that the AHC/2.1 software perhaps has defaults that can cause this behavior. I'm seeing a lot of different users that are only sending a few HTTP request to their Webhook.site endpoint, but have thousands of connections.  I don't think that they're knowlingly DDoSing Webhook.site.

While this can perhaps be remediated with better configuration/timeouts on my side, there seems to be some behavior in this software that is causing this issue.

**Comment by slandelle** [comment]
> One thing I'm noting that this issue is specific to the AHC/2.1 software. I'm not seeing this issue on any other HTTP clients.

Yeah, one bad behaving user is enough.
Also, AHC is pretty popular amongst web crawlers and those people tend to go through public proxies list to hide their deeds.

> While this can perhaps be remediated with better configuration/timeouts on my side, there seems to be some behavior in this software that is causing this issue.

Honestly, I don't think so. Even if these users don't really mean to harm you, they could be misusing this library. Many times I've seen people creating one client instance per request (first error) and not close it (second error).

Anyway, this project is basically dead. I've stopped maintaining years ago and the person who was supposed to take over never did the job. I just answered here because I felt sorry for you.
